docs/advanced.rst

```diff
@@ -243,3 +241,170 @@ the functions defined in their class body.
+2. The user is using the :func:`.chat_in_terminal` development utility
+3. The message was provided to an engine that does not support the class (e.g. an image part in a text-only model).
+
+The default behaviour is to transform the part to a Python-esque representation of its data and log a warning. You can
+override this method to specify the canonical string representation of your message part.
+
+.. code-block:: python
+
+    from kani import MessagePart
+
+    class ThoughtPart(MessagePart):
+        # hold some string data that is the chain-of-thought
+        data: str
+
+        # when stringified, this part should be hidden from the user/any engine that does not explicitly support it
+        def __str__(self):
+            return ""
+
+
+This will automatically register serialization and deserialization schemes for your class, so that all MessageParts
+are compatible with :meth:`.Kani.save` and :meth:`.Kani.load`.
+
+.. warning::
+    If you change the attributes or location of the class definition, old data may fail to be loaded into a new class.
+
+Using MessageParts in Engines
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Once we define a message part, we need to tell the engine how to use it. Since message parts are designed to provide
+engine-specific metadata while maintaining cross-engine compatibility, engines should check for the classes they know
+and cast unknown classes to a string. This lets parts control how they will be serialized in engines that don't natively
+support them.
+
+A common, but not necessary, pattern is to write an engine that *wraps* other engines. This wrapper engine acts as a
+translation layer for specific message parts, eventually compiling a single string (or list of strings) for a base
+engine (e.g. LlamaEngine) to consume.
+
+Here's the implementation for our chain-of-thought example. Note how we translate the parts by building a new message
+with a string content composed of parts in ``translate_message``, then use that method in our implementations of
+``message_len`` and ``predict``.
+
+.. code-block:: python
+
+    from kani import AIFunction, ChatMessage, MessagePart
+    from kani.engines.base import BaseEngine, Completion
+
+    class ChainOfThoughtEngine(BaseEngine):
+        def __init__(self, engine: BaseEngine):
+            self.engine = engine
+            self.max_context_size = engine.max_context_size
+
+        @staticmethod
+        def translate_message(message: ChatMessage) -> ChatMessage:
+            """Translate a input message into a simple string-only message to pass to the underlying engine."""
+            content = ""
+            # iterate over parts: if it's a thought, place its data in the content; otherwise prefix it with "Answer: "
+            for part in message.parts:
+                if isinstance(part, ThoughtPart):
+                    content += f"{part.data}\n"
+                else:
+                    content += f"Answer: {str(part)}"
```

Comment:
```
parts = [f"{p.data}\n" if isinstance(p, ThoughtPart) else f"Answer: {p}" for p in message.parts]
content = ''.join(parts)
```
Feels cleaner to me, but up to you
Also I feel like default behavior for f-strings is to call `__str__` implicitly so I think we can just do `f"Answer: {part}"`